### ISSUE TITLE
Directory component missing in child process program path argument

### ISSUE THREAD

**Issue by sivachandran** [civetweb/civetweb#555]
CivetWeb passes child process program path as first argument(argv[0]) to the child process. But the program path has no directory component in it. It is relative to the working directory(which is derived from the URI) but the relative directory specification `./` is missing in the path. For example, if the child program path is `/www-root/dir1/program1.exe` then the child program is spawned with `/www-root/dir1/` as working directory and `program1.exe` first argument to the child process. Whereas the first argument should be `./program1.exe` to be a valid relative path. The related source line is https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L10278.

The above behaviour prevents certain applications to be executed as CGI programs. For example a Python program converted as executable using cx_Freeze uses process's first argument to determine under which directory the program is running. The Python application fails to launch with error `OSError: No such file or directory ''` as the directory component is empty/blank from the process's first argument. As temporary fix I passed the absolute path as process's first argument and the Python application executed fine. A proper fix is to construct `p` in https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L10278 with relative path specification `./`.

**Comment by sivachandran** [comment]
A side observation, the cx_Freeze frozen Python application executes fine under Windows build of CivetWeb.

**Comment by bel2125** [comment]
I never worked with cx_Freeze - but I wonder there is a difference between `program1.exe` and `./program1.exe`. By default, all path should be relative, unless they start with `/` (Linux) or `\`, `\\` or `X:\` (Windows).
Did you configure [`cgi_interpreter`](https://github.com/civetweb/civetweb/blob/master/docs/UserManual.md#cgi_interpreter) or did you take the default value?

**Comment by bel2125** [comment]
Could you provide some steps to reproduce this problem?

**Comment by sivachandran** [comment]
@bel2125 
> By default, all path should be relative, ...
I don't think this is true. By default it checks whether the program exist in any of the path specified in `PATH` environment variable. Please refer http://www.linfo.org/dot_slash.html and the SO question https://stackoverflow.com/a/6331085/216534. So prefix `./` is required if you are referring a path(i.e. program) relative to the current directory.

I didn't configure `cgi_interpreter` as it doesn't require the interpreter.

I can definitely create a sample Python executable frozen by cx_Freeze but I'm not sure whether it would execute fine on your Linux distribution. I'm doubt about the dependency on C runtime libraries and probable incompatibilities between distribution. The other option to create self contained Docker image to demonstrate the issue. Both option will take some time as I'm busy with other things. Will try to send a sample app as soon as possible.

**Comment by bel2125** [comment]
I can run a test in Windows as well.

Just to be sure:
You create a Windows executable file with cx_Freeze.
Then this executable is started directly from civetweb, without a CGI interpreter.
You do not need an additional command line parameter - just this script.

Then it makes a difference, if it is started as "program.exe" or "./program.exe"?
Does it also make a difference, if you start it from cmd?
Window usually uses "\" instead of "/" - what about ".\program.exe"?

**Comment by sivachandran** [comment]
As I mentioned earlier the program works fine under Windows build of CivetWeb. Maybe the way child process is spawned(CreateProcess vs fork, execle) changes the behaviour. So to reproduce the issue you need to run the program in Linux.

In command-line you can't launch a program from current directory until you prefix with `./` except when the current directory is part of `PATH`.

**Comment by bel2125** [comment]
So, you start a program called "program.exe" in Linux.
If you call an executable file `.exe` in Linux, you need to add `**.exe$` to the `cgi_pattern` configuration.
The default is `cgi_pattern **.cgi$|**.pl$|**.php$`
Or better, you rename the file from `program.exe` to `program.cgi` and keep the default pattern.

**Comment by bel2125** [comment]
Within the [`test` directory](https://github.com/civetweb/civetweb/tree/master/test) there is a file [`linux.cgi`](https://github.com/civetweb/civetweb/blob/master/test/linux.cgi). If you start the server in Linux with `./civetweb -document_root test` the `linux.cgi` file will be executed when you open http://localhost:8080/test in a browser - please try this yourself.
If you put your executable in place of `linux.cgi`, it should be executed as well when you open http://localhost:8080/test in a browser (and hit reload) - please try this as well.

If both works, I did not understand the problem. In this case, please post the result of `./civetweb -I` for the system where the problem can be reproduced. Please post the complete command line used to start civetweb and the civetweb.conf file. I will also need a complete list of detailed steps to reproduce the problem.

**Comment by sivachandran** [comment]
I changed the `cgi_pattern` as `**.exe$` during my testing. Actually the program also getting launched. But the problem is the way the arguments are passed to the CGI programs. In fact the problem I'm reporting wouldn't affect any program if they don't use the first argument/arg0(i.e. executable name). It is just that cx_Freeze happen to use the arg0 for determining the executable directory and failing on CivetWeb launched executables as it didn't carry the directory component.

I've been trying to get you a Docker image that reproduces the issue. But it is getting tricky as I'm not much familiar with cx_Freeze.

**Comment by bel2125** [comment]
I did a test with the following script (`linux.cgi`):
```
#!/bin/sh

printf "Content-Type: text/plain\r\n"
printf "\r\n"

echo "This is a shell script called by CGI:"
echo
echo "Arg0: $0"
```

If I put it in the test folder, and start ./civetweb -document_root test, then test with the browser, I get "Arg0: linux.cgi"

If I test it from the shell 
using test/linux.cgi I get "Arg0: test/linux.cgi" 
using ./test/linux.cgi I get "Arg0: ./test/linux.cgi"
and from the test directory (cd test)
using ./linux.cgi I get "Arg0: ./linux.cgi" 
using sh linux.cgi I get "Arg0: linux.cgi"

So, you cannot assume arg[0] starts with "./" anyway.
If cx_Freeze really requires arg[0] to start with "./", I think this is an issue there, since it will not even work from a standard Linux bash - independent from CivetWeb.

**Comment by sivachandran** [comment]
Agree, it could be a cx_Freeze issue too. But the same program works fine under Mongoose without any change. This is the reason I doubted the problem could be in CivetWeb than cx_Freeze.

**Comment by bel2125** [comment]
Did you test your script in a sub-directory as well?
While the linux shell will not execute "linux.cgi" (only "./linux.cgi" or "sh linux.cgi" or ". linux.cgi" unless you are in a special path), it will definitely execute "test/linux.cgi".
Could you post a simple python program (not frozen) to reproduce this?

**Comment by sivachandran** [comment]
From the shell I tried running it under sub-directory and it works.

The python script is as simple as the below one
```python
#!/usr/bin/python
import os
print("Content-Type: text/plain \n\n")
print("This is from python")
```

**Comment by bel2125** [comment]
I just added your script to the test folder ([pytest.linux.cgi](https://github.com/civetweb/civetweb/blob/master/test/pytest.linux.cgi), e16c39cc796ab4b2c408cc9ecb4c5bb0c1f70ad3) - it works fine, both directly in the document\_root directory and in a sub-directory.

I tested the "not frozen" script.
I don't have cx_Freeze installed, and "python -m pip install cx_Freeze --upgrade" (from https://anthony-tuininga.github.io/cx_Freeze/) does not work on my Linux PC ("/usr/bin/python: No module named pip").
Anyway, since it works in plain python and apparently does not work in cx_freeze, this is yet another indication the problem is in cx_freeze, not civetweb.

**Comment by sanyaade-iot** [comment]
Hi Everybody, 

I am trying to compile civetweb for ESP8266 microcontroller and to be able to call a CGI executable without the use of shebang inside top most of my scripts. Is there a civetweb example code to achieve this? If so in which file do I place the code? Many thanks!

I went through the document and follow all the above discussion but I am still lost.
For example in Mongoose if I declare a --> **static struct mg_serve_http_    opts s_http_server_opts;**  on top of my own C implementation file below the line  **#include mongoose.h**.  Then I can use the code below to add directive for my custom CGI executable: 

`// Set up HTTP server parameters
  mg_set_protocol_http_websocket(nc);
  s_http_server_opts.document_root = "web_root";  // Set up web root directory
  if (mg_stat(s_http_server_opts.document_root, &st) != 0) {
    fprintf(stderr, "%s", "Cannot find web_root directory, exiting\n");
    exit(1);
	}
    
 **s_http_server_opts.cgi_interpreter = "fs/ph7_cgi";**	//set up cgi-interpreter directory
  **s_http_server_opts.cgi_file_pattern = "**.php$";**
  **s_http_server_opts.index_files = "index.php, index.html, index.htm, index.shtml, index.cgi";**
  if (mg_stat(s_http_server_opts.cgi_interpreter, &cgi_st) != 0) {
    fprintf(stderr, "%s", "Cannot find cgi_interpreter directory, exiting\n");
    exit(1);
  }
The bold lines in above code setup my cgi_interpreter executable and directory path; the file pattern and index files.
I could also do: -->   **s_http_server_opts.cgi_interpreter = "./ph7_cgi";**	//set up cgi-interpreter directory.

As I porting to a ROM / Flash memory based device, everything run inside the hardware, no shell or console,and no commandline arguments to be run or passed to the device. The whole program is flashed into ESP8266Ex (a 4MB capacity microcontroller device).

How can I achieve the above equivalent with civetweb? Many thanks

Hear from you soon!

God blesses!!!

Best regards,
Sanyaade

**Comment by bel2125** [comment]
Currently I can't see any relation between the "ESP8266" issue (https://github.com/civetweb/civetweb/issues/555#issuecomment-388175875) and the original "cx_freeze" issue.

Let's treat the "ESP8266" issue here: #617

Are there any new insights on "cx_freeze" in the meantime?

**Comment by bel2125** [comment]
No activity since 2 years, ... closing this issue.
If there is still a problem with "cx_freeze" that needs to be solved in CivetWeb 1.12, please create a new issue.
